redis7/redis.go

```diff
@@ -0,0 +1,353 @@
+		cmdErrs, err := pl.Exec(context.TODO())
+		if err != nil {
+			for _, cmdErr := range cmdErrs {
+				s.log.WithError(cmdErr.Err()).Errorf("Redis pipeline command failed: %s", cmdErr.String())
+			}
+			opts.InvokeOnRollback()
+			return stoabs.ErrCommitFailed
+		}
+		opts.InvokeAfterCommit()
+	} else {
+		s.log.WithError(appError).Warn("Rolling back transaction due to application error")
+		pl.Discard()
+		opts.InvokeOnRollback()
+		return appError
+	}
+
+	return nil
+}
+
+func (s *store) checkOpen() error {
+	s.mux.RLock()
+	defer s.mux.RUnlock()
+	if s.client == nil {
+		return stoabs.ErrStoreIsClosed
+	}
+	return nil
+}
+
+type tx struct {
+	reader redis.Cmdable
+	writer redis.Cmdable
+	store  *store
+}
+
+func (t tx) GetShelfWriter(shelfName string) (stoabs.Writer, error) {
+	return t.store.getShelf(shelfName, t.writer, t.reader), nil
+}
+
+func (t tx) GetShelfReader(shelfName string) (stoabs.Reader, error) {
+	reader, err := t.store.getShelfReader(shelfName)
+	if err != nil {
+		return nil, err
+	}
+	if reader == nil {
+		return nil, nil
+	}
+	return reader, nil
+}
+
+func (t tx) Store() stoabs.KVStore {
+	return t.store
+}
+
+func (t tx) Unwrap() interface{} {
+	return t.writer
+}
+
+type shelf struct {
+	name   string
+	prefix string
+	reader redis.Cmdable
+	// writer holds the Redis command writer for writing to the database.
+	// It is separate from the reader because writes use a transactional Redis pipeline, which buffers the commands.
+	// Buffering read commands would result in unexpected behaviour where the reads are only executed on transaction commit.
+	writer redis.Cmdable
+	store  *store
+}
+
+func (s shelf) Put(key stoabs.Key, value []byte) error {
+	if err := s.store.checkOpen(); err != nil {
+		return err
+	}
+	return s.writer.Set(context.TODO(), s.toRedisKey(key), value, 0).Err()
+}
+
+func (s shelf) Delete(key stoabs.Key) error {
+	return s.writer.Del(context.TODO(), s.toRedisKey(key)).Err()
+}
+
+func (s shelf) Get(key stoabs.Key) ([]byte, error) {
+	result, err := s.reader.Get(context.TODO(), s.toRedisKey(key)).Result()
+	if err == redis.Nil {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+	return []byte(result), nil
+}
+
+func (s shelf) Iterate(callback stoabs.CallerFn) error {
+	var cursor uint64
+	var err error
+	var keys []string
+	for {
+		scanCmd := s.reader.Scan(context.TODO(), cursor, s.toRedisKey(stoabs.BytesKey("*")), resultCount)
+		keys, cursor, err = scanCmd.Result()
+		if err != nil {
+			return err
+		}
+		err := s.visitKeys(keys, callback)
+		if err != nil {
+			return err
+		}
+		if cursor == 0 {
+			// Done
+			break
+		}
+	}
+	return nil
+}
+
+func (s shelf) Range(from stoabs.Key, to stoabs.Key, callback stoabs.CallerFn) error {
+	keys := make([]string, 0, resultCount)
+	// Iterate from..to (start inclusive, end exclusive)
+	var numKeys = 0
+	for curr := from; bytes.Compare(curr.Bytes(), to.Bytes()) == -1; curr = curr.Next() {
+		keys = append(keys, s.toRedisKey(curr))
```

Comment:
Just append the keys, no transformation

Reply:
They need to be in fully qualified format (shelf:key), because otherwise visitKeys won't find them

Reply:
Then Iterate() will have to convert its Redis keys to stoabs.Key and visitKeys will convert it back to Redis keys, so it will only add extra convertion steps.